agent/dns_test.go

```diff
@@ -4591,99 +4593,76 @@ func TestDNS_ServiceLookup_TTL(t *testing.T) {
-	}
-
-	var out struct{}
-	if err := a.RPC("Catalog.Register", args, &out); err != nil {
-		t.Fatalf("err: %v", err)
-	}
+	for idx, service := range []string{"db", "dblb", "dk", "api"} {
+		nodeName := fmt.Sprintf("foo%d", idx)
+		address := fmt.Sprintf("127.0.0.%d", idx)
+		args := &structs.RegisterRequest{
+			Datacenter: "dc1",
+			Node:       nodeName,
+			Address:    address,
+			Service: &structs.NodeService{
+				Service: service,
+				Tags:    []string{"master"},
+				Port:    12345 + idx,
+			},
+		}
 
-	args = &structs.RegisterRequest{
-		Datacenter: "dc1",
-		Node:       "foo",
-		Address:    "127.0.0.1",
-		Service: &structs.NodeService{
-			Service: "api",
-			Port:    2222,
-		},
-	}
-	if err := a.RPC("Catalog.Register", args, &out); err != nil {
-		t.Fatalf("err: %v", err)
+		var out struct{}
+		if err := a.RPC("Catalog.Register", args, &out); err != nil {
+			t.Fatalf("err: %v", err)
+		}
 	}
 
-	m := new(dns.Msg)
-	m.SetQuestion("db.service.consul.", dns.TypeSRV)
-
 	c := new(dns.Client)
-	in, _, err := c.Exchange(m, a.DNSAddr())
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
-
-	if len(in.Answer) != 1 {
-		t.Fatalf("Bad: %#v", in)
-	}
-
-	srvRec, ok := in.Answer[0].(*dns.SRV)
-	if !ok {
-		t.Fatalf("Bad: %#v", in.Answer[0])
-	}
-	if srvRec.Hdr.Ttl != 10 {
-		t.Fatalf("Bad: %#v", in.Answer[0])
-	}
-
-	aRec, ok := in.Extra[0].(*dns.A)
-	if !ok {
-		t.Fatalf("Bad: %#v", in.Extra[0])
-	}
-	if aRec.Hdr.Ttl != 10 {
-		t.Fatalf("Bad: %#v", in.Extra[0])
-	}
+	expectResult := func(dnsQuery string, expectedTTL uint32) {
+		t.Run(dnsQuery, func(t *testing.T) {
+			m := new(dns.Msg)
+			m.SetQuestion(dnsQuery, dns.TypeSRV)
 
-	m = new(dns.Msg)
-	m.SetQuestion("api.service.consul.", dns.TypeSRV)
-	in, _, err = c.Exchange(m, a.DNSAddr())
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
+			in, _, err := c.Exchange(m, a.DNSAddr())
+			if err != nil {
+				t.Fatalf("err: %v", err)
+			}
 
-	if len(in.Answer) != 1 {
-		t.Fatalf("Bad: %#v", in)
-	}
+			if len(in.Answer) != 1 {
+				t.Fatalf("Bad: %#v, len is %d", in, len(in.Answer))
+			}
 
-	srvRec, ok = in.Answer[0].(*dns.SRV)
-	if !ok {
-		t.Fatalf("Bad: %#v", in.Answer[0])
-	}
-	if srvRec.Hdr.Ttl != 5 {
-		t.Fatalf("Bad: %#v", in.Answer[0])
-	}
+			srvRec, ok := in.Answer[0].(*dns.SRV)
+			if !ok {
+				t.Fatalf("Bad: %#v", in.Answer[0])
+			}
+			if srvRec.Hdr.Ttl != expectedTTL {
+				t.Fatalf("Bad: %#v", in.Answer[0])
+			}
 
-	aRec, ok = in.Extra[0].(*dns.A)
-	if !ok {
-		t.Fatalf("Bad: %#v", in.Extra[0])
-	}
-	if aRec.Hdr.Ttl != 5 {
-		t.Fatalf("Bad: %#v", in.Extra[0])
+			aRec, ok := in.Extra[0].(*dns.A)
+			if !ok {
+				t.Fatalf("Bad: %#v", in.Extra[0])
+			}
+			if aRec.Hdr.Ttl != expectedTTL {
+				t.Fatalf("Bad: %#v", in.Extra[0])
+			}
+		})
 	}
+	// Should have its exact TTL
+	expectResult("db.service.consul.", 10)
+	// Should match db*
+	expectResult("dblb.service.consul.", 66)
+	// Should match d*
+	expectResult("dk.service.consul.", 42)
+	// Should be the default value
```

Comment:
nit: this is a little confusing since the default is `0`, I'd say "Should match `*`"